### PR TITLE
fixes/c-oc-java (2018 edition)

### PR DESCRIPTION
**Pull request by aet**
New iteration of #6 with all changes re-applied manually against the latest master

Other than fixing random issues that I noticed against C/C++/Objective-C/Java sources, few other notable things:

- Tries to make language keyword maintenance a bit less error prone through inheritance
- C header files considered a new "language", .h usually can be a mixture of whatever C/C++/Objective-C
- Objective-C support modernization passes

Tested against 3 non-unity core source trees (Objective-C / Java) and fixed new regressions that have come since #6

No regressions against core unity source tree, fixes few formatting things

Few commits starting with (FIXED?) commit message seem to be fixed already in the upstream with some alternative way, but since the actual code changes still applied, I added them along the testcase.

The only code change entirely dropped from this PR is 01b65ac229d69f12a770dc5ba56214a0964b800b but the actual testcase is still included in 4c9ed33 to make sure it won't break later

### CODE REVIEW

File: src/keywords.cpp [hunk 1]
```diff
@@ -47,287 +47,342 @@ static int kw_compare(const void *p1, const void *p2);
...
+ *
+ * For potential special cases, it's still possible to override
+ * the type inheritance for both LANG_OC and LANG_CPP separately,
+ * where as with LANG_ALLC you don't have a choice for override
+ *
+ * Certain keywords with LANG_C are obviously not really a part
+ * of C, but since they have been previously defined for LANG_C,
+ * let's not break existing behaviour
+ *
+ * Type inheritance is implemented at init_keywords()
```
Comment (mihaipopescu) [inline src/keywords.cpp:69]: 💯

File: src/punctuators.cpp [hunk 2]
```diff
@@ -26,97 +26,97 @@ using namespace std;
...
+   { "%:@",      CT_POUND,        LANG_C | LANG_CPP | LANG_OC                                            }, // digraph  #@ MS extension
+   { R"_(??=)_", CT_POUND,        LANG_C | LANG_CPP | LANG_OC | FLAG_DIG                                 }, // trigraph #
+   { R"_(??()_", CT_SQUARE_OPEN,  LANG_C | LANG_CPP | LANG_OC | FLAG_DIG                                 }, // trigraph [
+   { R"_(??))_", CT_SQUARE_CLOSE, LANG_C | LANG_CPP | LANG_OC | FLAG_DIG                                 }, // trigraph ]
+   { R"_(??')_", CT_CARET,        LANG_C | LANG_CPP | LANG_OC | FLAG_DIG                                 }, // trigraph ^
+   { R"_(??<)_", CT_BRACE_OPEN,   LANG_C | LANG_CPP | LANG_OC | FLAG_DIG                                 }, // trigraph {
+   { R"_(??>)_", CT_BRACE_CLOSE,  LANG_C | LANG_CPP | LANG_OC | FLAG_DIG                                 }, // trigraph }
+   { R"_(??-)_", CT_INV,          LANG_C | LANG_CPP | LANG_OC | FLAG_DIG                                 }, // trigraph ~
+   { R"_(??!)_", CT_ARITH,        LANG_C | LANG_CPP | LANG_OC | FLAG_DIG                                 }, // trigraph |
```
Comment (mihaipopescu) [inline src/punctuators.cpp]: could've used the same language keywords inheritance concept here as well but these are fewer.

File: src/uncrustify.cpp [hunk 3]
```diff
@@ -2099,17 +2101,18 @@ struct lang_name_t
...
+   { "D",        LANG_D                        },
+   { "CS",       LANG_CS                       },
+   { "VALA",     LANG_VALA                     },
+   { "JAVA",     LANG_JAVA                     },
+   { "PAWN",     LANG_PAWN                     },
+   { "OC",       LANG_OC                       },
+   { "OC+",      LANG_OC | LANG_CPP            },
+   { "CS+",      LANG_CS | LANG_CPP            },
+   { "ECMA",     LANG_ECMA                     },
+   { "C-Header", LANG_OC | LANG_CPP | FLAG_HDR },
```
Comment (mihaipopescu) [inline src/uncrustify.cpp:2115]: Awesome 💯. We actually had a hack in our code to treat all header files as `OC+` as there was no way to tell them apart. I think this change needs to be documented in the PR description. cc @Rassi

**Review by mihaipopescu — APPROVED** [review]
I really welcome this long waited PR.

Along with all Objective-C fixes, it introduces two novelties that I like:
1. language keyword inheritance
2. new "language" for c header files

File: src/punctuators.cpp (hunk 2 above)
Comment (aet) [reply src/punctuators.cpp]: True, I also noticed that I had added a duplicate entry for LANG_OC into CT_ELLIPSIS, fixed in a95b4e8b

File: src/space.cpp [hunk 4]
```diff
@@ -843,8 +843,11 @@ static argval_t do_space(chunk_t *first, chunk_t *second, int &min_sp, bool comp
          log_rule("sp_template_angle");
          return(cpd.settings[UO_sp_template_angle].a);
       }
-      log_rule("sp_before_angle");
-      return(cpd.settings[UO_sp_before_angle].a);
+      if (first->type != CT_QUALIFIER)
+      {
+         log_rule("sp_before_angle");
+         return(cpd.settings[UO_sp_before_angle].a);
```
Comment (CDanU) [inline src/space.cpp:862]: Why is this if check necessary?

Comment (mihaipopescu) [reply src/space.cpp:862]: Is the change accompanied by a test?

Comment (CDanU) [reply src/space.cpp:862]: Yes, 018e14b9e24e9f5d8bb88b0d17ec83fa13db299d .. well kind of.

I am not sure why the option shouldn't also be applied if there is a `CT_QUALIFIER` before the generic `<`. Java should still compile if `remove` is used to generate something like `public static<T> void` ...

**Review by aet — COMMENTED** [review]
commit  9cc01c8 has a wrong commit id in the message, it should be e1e449f34

the issue tracking number is correct though

**Comment by mihaipopescu** [comment]
jenkins builder: test this please

**Pull request merged**